Minor additions to `algorithmic_differentiation.md` (#457)
compintell/Mooncake.jl: docs/src/understanding_mooncake/algorithmic_differentiation.md

# Directional Derivatives and Gradients

Now we turn to using forwards- and reverse-mode AD to compute the gradient of a function.

Recall that if ``D f[x] : \mathcal{X} \to \mathbb{R}`` is the Frechet derivative discussed here then ``D f[x](\dot{x})`` is the _directional derivative_ in the ``\dot{x}`` direction.

The _gradient_ of ``f : \mathcal{X} \to \mathbb{R}`` at ``x`` is defined to be the vector ``\nabla f (x) \in \mathcal{X}`` such that
```math
\langle \nabla f (x), \dot{x} \rangle = D f[x](\dot{x})
```
for any direction ``\dot{x}``.
In other words, the vector ``\nabla f`` encodes all the information about the directional derivatives of ``f``, and we use the inner product to retrieve each one.

> An alternative characterisation is that ``\nabla f(x)`` is the vector pointing in the direction of steepest ascent on ``f`` at ``x``, with magnitude equal to the directional derivative in that steepest direction.

Reviewer: I think I'm not quite clear what is meant by `with magnitude equal to the directional derivative in that steepest direction.` -- is there a precise mathematical statement by which you can explain what this means?

_**Aside: The choice of inner product**_

Notice that the value of the gradient depends on how the inner product on ``\mathcal{X}`` is defined.
Indeed, different choices of inner product result in different values of ``\nabla f``.
Adjoints such as ``D f[x]^*`` are also inner product dependent.

> However, the actual derivative ``D f[x]`` is of course invariant -- it makes no reference to the inner product.

Reviewer: Is this correct, technically? We make use of the norms for both X and Y in the definition of the Frechet derivative, which I've been assuming we take to be the norms induced by whichever inner products we pick on X and Y. Would it be more accurate to point out that the definition is invariant because all norms in are equivalent in finite dimensions?

> In practice, Mooncake uses the Euclidean inner product, extended in the "obvious way" to other composite data types (that is, as if everything is flattened and embedded in ``\mathbb{R}^N``).
> But we endeavour to keep the discussion general in order to make the role of the inner product explicit.

Reviewer: ```suggestion
In practice, Mooncake uses the Euclidean inner product, extended in the "obvious way" to other composite data types (that is, as if everything is flattened and embedded in ``\mathbb{R}^N``), but we endeavour to keep the discussion general in order to make the role of the inner product explicit.
```
grammar

#### Computing the gradient from forwards-mode

To compute the gradient in forwards-mode, we need to evaluate the forwards pass ``\dim \mathcal{X}`` times.
We also need to refer to a basis ``\{\mathbf{e}_i\}`` of ``\mathcal{X}`` and its reciprocal basis ``\{\mathbf{e}^i\}`` defined by ``\langle \mathbf{e}_i, \mathbf{e}^j \rangle = \delta_i^j``.

> (For any basis there exists such a reciprocal basis, and they are the same if the basis is orthonormal.)

Reviewer: ```suggestion
For any basis there exists such a reciprocal basis, and they are the same for orthonormal bases such as the standard basis. As a result, you can replace any occurrences of ``\{\mathbf{e}^i\}`` with ``\{\mathbf{e}_i\}`` in what follows and still have a correct understanding of the mathematics underpinning Mooncake.
```
nip-pick: I don't think we need the brackets here, and I think it would be good to allude to the standard basis.

What are your thoughts on my bit about "replace occurences of..."? I'm not 100% certain I've phrased this perfectly, but I would like to reassure readers that this is indeed the consequence of the previous sentence. Maybe I'm overthinking it...

Equipped with such a pair of bases, we can always decompose a vector ``x = \sum_i x^i \mathbf{e}_i`` into its components ``x^i = \langle x, \mathbf{e}^i \rangle``.
Therefore, the gradient is given by
```math
\nabla f(x)
	= \sum_i \langle \nabla f(x), \mathbf{e}^i \rangle \mathbf{e}_i
	= \sum_i D f[x](\mathbf{e}^i) \, \mathbf{e}_i
```

> where the second equality follows from the gradient's implicit definition.

Reviewer: ```suggestion
where the second equality follows from the gradient's definition.
```
Reading this, I briefly thought that we had multiple definitions of the gradient lying around, and the one you are using here is the "implicit" one, before realising you're just trying to point out that our definition of the gradient is implicit. I wonder if others might read it in the same way, meaning that it's better just to refer to the "gradient's definition"?

```math
\mathbf{e}_1 = \mathbf{e}^1 = (1, (0, 0)), \quad
\mathbf{e}_2 = \mathbf{e}^2 = (0, (1, 0)), \quad
\mathbf{e}_3 = \mathbf{e}^3 = (0, (0, 1)), \quad
```
so the gradient is
```math
\begin{align*}
\nabla f(x, y)
	&= \sum_i D f[x, y](\mathbf{e}^i) \mathbf{e}_i \\
	&= \Big(D f[x, y](1, (0, 0)), \big(D f[x, y](0, (1, 0)), D f[x, y](0, (0, 1))\big)\Big) \\
	&= (1, (y_2, y_1))
\end{align*}
```
referring [above](#AD-of-a-Julia-function:-a-slightly-less-trivial-example) for the form of ``D f[x, y]``.

#### Computing the gradient from reverse-mode
If we perform a single reverse-pass on a function ``f : \mathcal{X} \to \RR`` to obtain ``D f[x]^\ast``, then the gradient is simply
```math
\nabla f (x) = D f[x]^\ast (1) .
```

To show this, note that ``D f [x] (\dot{x}) = \langle 1, D f[x] (\dot{x}) \rangle = \langle D f[x]^\ast (1), \dot{x} \rangle`` using the definition of the adjoint.
Then, the definition of the gradient gives
```math
\langle \nabla f (x), \dot{x} \rangle = \langle D f[x]^\ast (1), \dot{x} \rangle
```
which implies ``\nabla f (x) = D f[x]^\ast (1)`` since ``\dot{x}`` is arbitrary.

_**Example**_

> The adjoint derivative of ``f(x, y) = x + y_1 y_2`` (see [above](#AD-of-a-Julia-function:-a-slightly-less-trivial-example)) immediately gives

Reviewer: ```suggestion
The adjoint of the derivative of ``f(x, y) = x + y_1 y_2`` (see [above](#AD-of-a-Julia-function:-a-slightly-less-trivial-example)) immediately gives
```
nit-pick: I don't _believe_ we refer to the "adjoint derivative" anywhere, but we do refer to the "adjoint" and the "adjoint of the derivative" interchangeably. Is this a typo, or ought we to be talking about the "adjoint derivative"?

```math
\nabla f(x, y) = D f[x, y]^\ast (1) = (1, (y_2, y_1)) .
```

> _**Aside: Adjoint Derivatives as Gradients**_

Reviewer: Same here -- "Adjoint Derivatives" vs "Adjoint" or "Adjoint of the Derivative" etc
